// typedoc/dist/lib/utils/options/readers/tsconfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TSConfigReader = void 0;
const path_1 = require("path");
const fs_1 = require("fs");
const ts = require("typescript");
const fs_2 = require("../../fs");
function isFile(file) {
    return fs_1.existsSync(file) && fs_1.statSync(file).isFile();
}
class TSConfigReader {
    constructor() {
        /**
         * Note: Runs after the [[TypeDocReader]].
         */
        this.priority = 200;
        this.name = "tsconfig-json";
    }
    read(container, logger) {
        const tsconfigOpt = container.getValue("tsconfig");
        if (!container.isDefault("tsconfig")) {
            this._tryReadOptions(tsconfigOpt, container, logger);
            return;
        }
        // Don't log errors if we try to read by default.
        this._tryReadOptions(tsconfigOpt, container);
    }
    _tryReadOptions(file, container, logger) {
        var _a, _b;
        let fileToRead = file;
        if (!isFile(fileToRead)) {
            fileToRead = ts.findConfigFile(file, isFile, file.toLowerCase().endsWith(".json")
                ? path_1.basename(file)
                : undefined);
        }
        if (!fileToRead || !isFile(fileToRead)) {
            logger === null || logger === void 0 ? void 0 : logger.error(`The tsconfig file ${file} does not exist`);
            return;
        }
        fileToRead = fs_2.normalizePath(path_1.resolve(fileToRead));
        let fatalError = false;
        const parsed = ts.getParsedCommandLineOfConfigFile(fileToRead, {}, {
            ...ts.sys,
            onUnRecoverableConfigFileDiagnostic(error) {
                logger === null || logger === void 0 ? void 0 : logger.diagnostic(error);
                fatalError = true;
            },
        });
        if (!parsed || fatalError) {
            return;
        }
        logger === null || logger === void 0 ? void 0 : logger.diagnostics(parsed.errors);
        const typedocOptions = (_b = (_a = parsed.raw) === null || _a === void 0 ? void 0 : _a.typedocOptions) !== null && _b !== void 0 ? _b : {};
        if (typedocOptions.options) {
            logger === null || logger === void 0 ? void 0 : logger.error([
                "typedocOptions in tsconfig file specifies an option file to read but the option",
                "file has already been read. This is likely a misconfiguration.",
            ].join(" "));
            delete typedocOptions.options;
        }
        if (typedocOptions.tsconfig) {
            logger === null || logger === void 0 ? void 0 : logger.error("typedocOptions in tsconfig file may not specify a tsconfig file to read");
            delete typedocOptions.tsconfig;
        }
        container.setCompilerOptions(parsed.fileNames, parsed.options, parsed.projectReferences);
        for (const [key, val] of Object.entries(typedocOptions || {})) {
            try {
                container.setValue(key, val);
            }
            catch (error) {
                logger === null || logger === void 0 ? void 0 : logger.error(error.message);
            }
        }
    }
}
exports.TSConfigReader = TSConfigReader;
